21.md

```diff
@@ -0,0 +1,173 @@
+	* 32 bits of random data is probably sufficient entropy to avoid collisions
+4. Sort the query string by key (alphabetically). The above example would result in the following:
+	* `amount=5&currency=EUR&id=935e30a7&nonce=d2e3c794&tag=withdraw`
+5. The sorted query string will be referred to as the "payload".
+	* Note that both the keys and values in the query string must be URL-encoded. The following characters should be __unescaped__: `A-Z a-z 0-9 - _ . ! ~ * ' ( )`. See [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent#description) for more details.
+6. Sign the payload using the authorization key secret. Signatures are generated using HMAC-SHA256, where the authorization key secret is the key and the payload is the message.
+7. Add the signature to the query string. Example:
+	* `amount=5&currency=EUR&id=935e30a7&nonce=d2e3c794&tag=withdraw&signature=80224eed83e03acd0e44760f42b3a7157f549d04cf0160574246e9a87ff9bf8f`
+8. Build the full URL with the signed query string. Example:
+	* https://example.com/lnurl?amount=5&currency=EUR&id=935e30a7&nonce=d2e3c794&tag=withdraw&signature=80224eed83e03acd0e44760f42b3a7157f549d04cf0160574246e9a87ff9bf8f
+
+
+### Steps to verify a signed URL
+
+These steps are to be done by the service when it receives a signed LNURL request:
+
+1. Given the following example request URL:
+	* https://example.com/lnurl?amount=5&currency=EUR&id=935e30a7&nonce=d2e3c794&tag=withdraw&signature=80224eed83e03acd0e44760f42b3a7157f549d04cf0160574246e9a87ff9bf8f
+2. Separate the query string from the rest of the request URL. Example:
+	* `amount=5&currency=EUR&id=935e30a7&nonce=d2e3c794&tag=withdraw&signature=80224eed83e03acd0e44760f42b3a7157f549d04cf0160574246e9a87ff9bf8f`
+2. Remove the signature from the query string. Example:
+	* `amount=5&currency=EUR&id=935e30a7&nonce=d2e3c794&tag=withdraw`
+3. Sort the query string by key (alphabetically) in the case that it is not already. Example:
+	* `amount=5&currency=EUR&id=935e30a7&nonce=d2e3c794&tag=withdraw`
+4. The sorted query string will be referred to as the "payload".
+5. Use the `id` in the query string to find the authorization key in the service's list of authorization keys.
+	* If the authorization key is __not__ found, then fail the request.
+	* If the authorization key is found, then continue with the signature check.
+6. Sign the payload using the authorization key secret that was found in the previous step. Signatures are generated using HMAC-SHA256, where the authorization key secret is the key and the payload is the message.
+7. Check the signature from the query string against the signature generated by the service.
+	* If the signature __does not__ match, then fail the request.
+	* If the signature matches, then continue with the service's normal LNURL flow.
+
+_Optional_ steps to generate a deterministic value to be used as the `k1` or identifier of an LNURL record:
```

Comment:
I don't think this can be optional. This makes LUD-21 incompatible with the base withdraw spec [LUD-03](https://github.com/fiatjaf/lnurl-rfc/blob/luds/03.md).

Specifically:

> 5. Once accepted by the user, `LN WALLET` sends a GET to `LN SERVICE` in the form of
> 	```
> 	<callback>
>       <?|&> // either '?' or '&' depending on whether there is a query string already in the callback
>        k1=<k1> // the k1 specified in the response above
>       &pr=<lightning invoice> // the payment request generated by the wallet
> 	```

Reply:
It's optional because you don't need to deterministically generate the `k1` from the signed URL. The service can randomly generate the `k1` and use that in the response object. This would mean that your service will accept any and all valid, signed LNURL and never check for re-use. Or you could find some other way to prevent re-use of signed LNURLs.